specs/TryUpdateRuntime.md

```diff
@@ -0,0 +1,229 @@
+    HRESULT hr = GetAvailableCoreWebView2BrowserVersionString(nullptr, &currentVersion);
+    if (FAILED(hr) || (currentVersion == nullptr))
+    {
+      return false;
+    }
+    int versionComparisonResult;
+    CompareBrowserVersions(currentVersion.get(), minimalVersionRequired.c_str(), &versionComparisonResult);
+    return (versionComparisonResult >= 0)
+}
+
+void EnsureWebView2RuntimeVersion(std::function<void(bool)> const& callback, std::wstring minimalVersionRequired)
+{
+    if (IsCurrentVersionSameOrNewer(minimalVersionRequired))
+    {
+      callback(true);
+      return;
+    }
+    auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
+    CreateCoreWebView2EnvironmentWithOptions(nullptr, nullptr, options.Get(),
+        Callback<ICoreWebView2CreateCoreWebView2EnvironmentCompletedHandler>(
+            [callback, minimalVersionRequired](HRESULT result, ICoreWebView2Environment* environment) -> HRESULT {
+                wil::com_ptr<ICoreWebView2Environment> webViewEnvironment = environment;
+                auto experimentalEnvironment3 =
+                webViewEnvironment.try_query<ICoreWebView2ExperimentalEnvironment3>();
+                HRESULT hr = experimentalEnvironment3->TryUpdateRuntime(
+                    Callback<ICoreWebView2ExperimentalTryUpdateRuntimeCompletedHandler>(
+                        [callback, minimalVersionRequired, experimentalEnvironment3](HRESULT errorCode,
+                           ICoreWebView2ExperimentalUpdateRuntimeResult* result) -> HRESULT {
+                            COREWEBVIEW2_RUNTIME_UPDATE_STATUS updateStatus =
+                                COREWEBVIEW2_RUNTIME_UPDATE_STATUS_FAILED;
+                            if ((errorCode == S_OK) && result)
+                            {
+                              CHECK_FAILURE(result->get_UpdateRuntimeStatus(&updateStatus));
+                            }
+                            if (updateStatus != COREWEBVIEW2_RUNTIME_UPDATE_STATUS_UPDATED)
+                            {
+                              callback(false);
+                            }
+                            else
+                            {
+                              callback(IsCurrentVersionSameOrNewer(minimalVersionRequired));
+                            }
+                            return S_OK;
+                        })
+                        .Get());
+                return S_OK;
+            })
+            .Get());
+}
+
+```
+
+# Remarks
+See comments in [API Details](#api-details) section below.
+
+# API Notes
+See [API Details](#api-details) section below for API reference.
+
+# API Details
+
+## Win32 C++
+```IDL
+/// Status of TryUpdateRuntime operation result.
+[v1_enum] typedef enum COREWEBVIEW2_UPDATE_RUNTIME_STATUS {
+
+  /// No update for Edge WebView2 Runtime is available.
+  /// Latest version of Edge WebView2 Runtime is already installed.
+  COREWEBVIEW2_UPDATE_RUNTIME_STATUS_NO_UPDATE,
```

Comment:
The fact that both the C# and C++ samples mishandled the "no update" case suggests that maybe we shouldn't distinguish the two success cases ("update needed and applies" and "update not needed") in the status code. In particular because race conditions render the "No update" case meaningless.

Possible names

* `COREWEBVIEW2_UPDATE_RUNTIME_STATUS_LATEST_VERSION_INSTALLED`
* `COREWEBVIEW2_UPDATE_RUNTIME_STATUS_BLOCKED_BY_POLICY`
* `COREWEBVIEW2_UPDATE_RUNTIME_STATUS_FAILED`

We can report the secondary status ("No update needed" vs "Update needed and applied") as a separate property, like we do UpdateError, but as noted, since there's a race condition involved, even that result is unreliable.

Reply:
That is true, even if we report a secondary status, that secondary status would not be reliable, so probably simply report with COREWEBVIEW2_UPDATE_RUNTIME_STATUS_LATEST_VERSION_INSTALLED and the app should always recheck using installed version.